### PR TITLE
Add lambda to store GitHub Actions status

### PR DESCRIPTION
**Pull request by driazati**
This lambda is intended to be run on a CloudWatch event that fires every few minutes or so, the lambda itself handles debouncing so it does not write data sooner than 5 minutes from the previous entry. It will append to a single `status.json` file the latest results from GitHub's workflows API. It crawls through pages of results from GitHub until it sees that all the jobs on that page are `completed` and records the counts of each into `status.json`.

### CODE REVIEW

File: fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            return
+
+        if len(all_stats) >= MAX_LEN:
+            # Chop off old data
+            all_stats = all_stats[:MAX_LEN]
+
+        all_stats.insert(0, await get_gha_statuses())
+        print("writing", all_stats)
+        await client.put_object(
+            Bucket=bucket_name, Key="status.json", Body=json.dumps(all_stats)
```
Comment (zhouzhuojie) [inline fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:132]: how fast can the status.json grow? or is it just cap at the latest max batch size?

**Review by malfet — APPROVED** [review]

File: .github/workflows/deploy_checks_lambda.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      - 'fbossci-aws-lambas/us-east-1/checks-cron/**'
+
+jobs:
+  deploy:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Deploy
+        uses: appleboy/lambda-action@1e05c1377056f21ebb2ce69b910bc16b943c2a66
```
Comment (walterddr) [inline .github/workflows/deploy_checks_lambda.yml:18]: qq: `timeout` defaulted to 3 sec according to https://github.com/appleboy/lambda-action#input-variables
did we profile the speed of this lambda

**Review by walterddr — APPROVED** [review]

**Review by seemethere — APPROVED** [review]

**Comment by samestep** [comment]
Just to make sure I understand: what are the `Makefile` and `requirements.txt` files for in `fbossci-aws-lambas/us-east-1/checks-cron/`, since `.github/workflows/deploy_checks_lambda.yml` only uses the `lambda_function.py` file in that folder?

File: fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        )
+
+
+def lambda_handler(event, context):
+    print("handling lambda")
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(main())
```
Comment (samestep) [inline fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:139]: Is there a particular reason for not just doing this instead?
```suggestion
    asyncio.run(main())
```

File: fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py (hunk 1 above)
Comment (samestep) [reply fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:132]: yeah that's what the `all_stats = all_stats[:MAX_LEN]` line does I believe

File: fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        content = await content["Body"].read()
+        all_stats = json.loads(content.decode())
+
+        if not should_check_github(all_stats):
+            print("Ran too early, not doing anything")
+            return
+
+        if len(all_stats) >= MAX_LEN:
+            # Chop off old data
+            all_stats = all_stats[:MAX_LEN]
```
Comment (samestep) [inline fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:127]: The `if` check is superfluous:
```suggestion
        # Chop off old data
        all_stats = all_stats[:MAX_LEN]
```

File: fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+        all_stats.insert(0, await get_gha_statuses())
```
Comment (samestep) [inline fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:129]: Wait, I'm confused... I thought the `should_check_github` function looked at the _last_ element of the list? If this `insert`s at the beginning, doesn't that bypass the 5-minute time check?

File: fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                all_statuses[run["status"]] += 1
+                new_statuses[run["status"]] += 1
+
+            if not page_in_progress(new_statuses):
+                pages_past -= 1
+            else:
+                pages_past = max_pages_past
+
+            if pages_past == 0:
+                should_quit = True
```
Comment (samestep) [inline fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:98]: Could you explain the reasoning behind this? I don't quite understand

Comment (driazati) [reply fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:98]: There's no way to get all the unfinished jobs from the GitHub Actions API, here this assumes that it returns the most recent data first (and it's all paginated via the API), the heuristic here is that once the lambda has seen a certain number of completed jobs with no pending/queued/in progress jobs in between, it's probably safe to assume we've seen all the unfinished jobs. The stuff here is just making sure it checks a certain number of completed jobs before moving on (page size * max_pages_past)

Comment (samestep) [reply fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:98]: Makes sense, thanks! Could you add a code comment saying this?

File: .github/workflows/deploy_checks_lambda.yml (hunk 2 above)
Comment (driazati) [reply .github/workflows/deploy_checks_lambda.yml:18]: I ended up setting it manually to 15 seconds on the UI so added that here

File: fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py (hunk 5 above)
Comment (driazati) [reply fbossci-aws-lambas/us-east-1/checks-cron/lambda_function.py:128]: good catch, yes it should be checking the beginning and chopping off old data at the end of the list

**Comment by driazati** [comment]
> Just to make sure I understand: what are the `Makefile` and `requirements.txt` files for in `fbossci-aws-lambas/us-east-1/checks-cron/`, since `.github/workflows/deploy_checks_lambda.yml` only uses the `lambda_function.py` file in that folder?

It was mostly cribbed from the existing https://github.com/pytorch/test-infra/blob/master/.github/workflows/deploy_lambda.yml but it's not really clear where `requirements.txt` comes in there either. The latest update changes it to use the Makefile to build a zip, then that gets uploaded with the action.

**Pull request merged**